### PR TITLE
Removes redundant buffer zeroing in PseudoTerminal.swift by using init(unsafeUninitializedCapacity:initializingWith:)

### PR DESCRIPTION
**Pull request by valeriyvan**

### CODE REVIEW

**Comment by aciidb0mb3r** [comment]
@swift-ci smoke test

**Comment by aciidb0mb3r** [comment]
Thanks! Can you also import this change to the swift-tools-support-core repository?

**Comment by aciidb0mb3r** [comment]
```
18:39:41 Test Case '-[TSCBasicTests.TerminalControllerTests testBasic]' started.
18:39:41 Exited with signal code 4
18:39:41 env SDKROOT=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk SWIFTCI_USE_LOCAL_DEPS=1 DYLD_LIBRARY_PATH=/Users/buildnode/jenkins/workspace/swift-package-manager-PR-osx-smoke-test/branch-master/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/bootstrap/lib:/Users/buildnode/jenkins/workspace/swift-package-manager-PR-osx-smoke-test/branch-master/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/tsc/lib:/Users/buildnode/jenkins/workspace/swift-package-manager-PR-osx-smoke-test/branch-master/buildbot_incremental/llbuild-macosx-x86_64/lib /Users/buildnode/jenkins/workspace/swift-package-manager-PR-osx-smoke-test/branch-master/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/release/swift-test --enable-test-discovery --parallel --build-path /Users/buildnode/jenkins/workspace/swift-package-manager-PR-osx-smoke-test/branch-master/buildbot_incremental/swiftpm-macosx-x86_64 --configuration release -Xlinker -rpath -Xlinker @executable_path/../lib/swift/macosx
18:39:41 
--- bootstrap: error: Command '['env', 'SDKROOT=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk', 'SWIFTCI_USE_LOCAL_DEPS=1', u'DYLD_LIBRARY_PATH=/Users/buildnode/jenkins/workspace/swift-package-manager-PR-osx-smoke-test/branch-master/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/bootstrap/lib:/Users/buildnode/jenkins/workspace/swift-package-manager-PR-osx-smoke-test/branch-master/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/tsc/lib:/Users/buildnode/jenkins/workspace/swift-package-manager-PR-osx-smoke-test/branch-master/buildbot_incremental/llbuild-macosx-x86_64/lib', u'/Users/buildnode/jenkins/workspace/swift-package-manager-PR-osx-smoke-test/branch-master/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/release/swift-test', '--enable-test-discovery', '--parallel', '--build-path', '/Users/buildnode/jenkins/workspace/swift-package-manager-PR-osx-smoke-test/branch-master/buildbot_incremental/swiftpm-macosx-x86_64', '--configuration', 'release', '-Xlinker', '-rpath', '-Xlinker', u'@executable_path/../lib/swift/macosx']' returned non-zero exit status 1
```

Failure seems related to this change.

**Comment by aciidb0mb3r** [comment]
@swift-ci smoke test

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
thanks @valeriyvan can you explain the motivation fo this change

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by valeriyvan** [comment]
@tomerd
> thanks @valeriyvan can you explain the motivation fo this change

Avoiding redundant buffer zeroing makes code faster.

**Comment by tomerd** [comment]
> Avoiding redundant buffer zeroing makes code faster.

thanks @valeriyvan to clarify the question: was there a perf issue you noticed, or is this stemming from an observation that the code can be optimized?

**Comment by valeriyvan** [comment]
No, I haven not seen performance issue here.

File: swift-tools-support-core/Sources/TSCTestSupport/PseudoTerminal.swift
Comment (abertelrud) [inline swift-tools-support-core/Sources/TSCTestSupport/PseudoTerminal.swift:34]: It would probably be good to create a buffer of `n + 1` instead of `n` here, and then read up to `n` instead of `n -1` bytes below.  It's a small nuance, and client's shouldn't count on receiving the maximum requested number of bytes anyway, but in this new implementation it would always receive at most one less than the requested maximum.

**Comment by abertelrud** [comment]
This seems to also fix a possible problem with the old implementation.  Maybe I'm reading this wrong, but it looks as if the old implementation could run off the end of the buffer in `String(cString:)` since it takes an unsafe pointer and not a buffer with a length, and there was no guarantee of null-termination in the old code.

File: swift-tools-support-core/Sources/TSCTestSupport/PseudoTerminal.swift
Comment (valeriyvan) [reply swift-tools-support-core/Sources/TSCTestSupport/PseudoTerminal.swift]: Agree. This change makes sense. Fixed.

File: swift-tools-support-core/Sources/TSCTestSupport/PseudoTerminal.swift
Comment (abertelrud) [inline swift-tools-support-core/Sources/TSCTestSupport/PseudoTerminal.swift:37]: Looks good now, the only remaining question is how happy `unsafeUninitializedCapacity:` is if you return a negative number by reference (as it would be if `read()` returned an error).  So maybe it should be set to zero before returning.  I'm not familiar enough with `unsafeUninitializedCapacity:` to know if it's allowed to be negative on return; the documentation says "Inside the closure, set the initializedCount parameter to the number of elements that are initialized by the closure." and doesn't address a negative count, so I assume it's disallowed.

Apart from that it looks good!

**Comment by abertelrud** [comment]
@swift-ci please test

File: swift-tools-support-core/Sources/TSCTestSupport/PseudoTerminal.swift
Comment (valeriyvan) [reply swift-tools-support-core/Sources/TSCTestSupport/PseudoTerminal.swift]: Good point! Fixed that as well.

**Comment by abertelrud** [comment]
Thanks, looks good!

**Comment by abertelrud** [comment]
@swift-ci please test

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Comment by abertelrud** [comment]
Will have to look closer at the tests, but on the face of it, the errors seem unrelated:
```
CMake Error at Sources/SwiftDriver/CMakeLists.txt:102:
  Parse error.  Function missing ending ")".  Instead found unterminated
  string with text "
```
Still, need to figure this out before merging.

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Comment by abertelrud** [comment]
@valeriyvan The latest failures are crashes in unit tests that look related to this:
```
Test Suite 'Selected tests' started at 2020-09-05 00:47:49.534
Test Suite 'ProgressAnimationTests' started at 2020-09-05 00:47:49.536
Test Case 'ProgressAnimationTests.testNinjaProgressAnimationTTY' started at 2020-09-05 00:47:49.536
Exited with signal code 11

Test Suite 'Selected tests' started at 2020-09-05 00:47:49.537
Test Suite 'ProgressAnimationTests' started at 2020-09-05 00:47:49.539
Test Case 'ProgressAnimationTests.testPercentProgressAnimationTTY' started at 2020-09-05 00:47:49.539
Exited with signal code 11

Test Suite 'Selected tests' started at 2020-09-05 00:47:50.174
Test Suite 'TerminalControllerTests' started at 2020-09-05 00:47:50.175
Test Case 'TerminalControllerTests.testBasic' started at 2020-09-05 00:47:50.175
Exited with signal code 11
```
Would you able to look into it?

File: swift-tools-support-core/Sources/TSCTestSupport/PseudoTerminal.swift
Comment (abertelrud) [inline swift-tools-support-core/Sources/TSCTestSupport/PseudoTerminal.swift:36]: This should actually be `initializedCapacity = read(master, buf.baseAddress, n)`, because otherwise you're actually overwriting the pointer itself.  The reason the code doesn't just copy over is that in the old code `buf` was an array, but in the new code `buf` is a `UnsafeMutableBufferPointer`.

**Review by abertelrud — CHANGES_REQUESTED** [review]

**Comment by abertelrud** [comment]
@swift-ci please test

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Review by abertelrud — APPROVED** [review]
Approved and ready to land when the tests finish.

**Comment by abertelrud** [comment]
Tests pass (including the full-stack tests); this looks good.  Thanks again for the contribution!

**Pull request merged**
